cmd/nitro/init.go

```diff
@@ -606,132 +606,112 @@ func rebuildLocalWasm(ctx context.Context, config *gethexec.Config, l2BlockChain
+	if err != nil {
+		return nil, nil, err
 	}
 
-	// Check if database was misplaced in parent dir
-	const errorFmt = "database was not found in %s, but it was found in %s (have you placed the database in the wrong directory?)"
-	parentDir := filepath.Dir(stack.InstanceDir())
-	if dirExists(path.Join(parentDir, "l2chaindata")) {
-		return nil, nil, fmt.Errorf(errorFmt, stack.InstanceDir(), parentDir)
-	}
-	grandParentDir := filepath.Dir(parentDir)
-	if dirExists(path.Join(grandParentDir, "l2chaindata")) {
-		return nil, nil, fmt.Errorf(errorFmt, stack.InstanceDir(), grandParentDir)
-	}
+	if executionDB == nil {
+		err := checkAndDownloadDB(ctx, stack, config)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	if err := checkEmptyDatabaseDir(stack.InstanceDir(), config.Init.Force); err != nil {
-		return nil, nil, err
-	}
+		executionDB, wasmDB, err = openDownloadedExecutionDB(stack, config, cacheConfig, persistentConfig)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	if err := setLatestSnapshotUrl(ctx, &config.Init, config.Chain.Name); err != nil {
-		return nil, nil, err
-	}
+		var initDataReader *statetransfer.InitDataReader
+		var genesisArbOSInit *params.ArbOSInit
 
-	initFile, cleanUpTmp, err := initializeAndDownloadInit(ctx, &config.Init, stack)
-	defer cleanUpTmp()
-	if err != nil {
-		return nil, nil, err
-	}
+		initDataReader, chainConfig, genesisArbOSInit, err = getInit(config)
+		if err != nil {
+			return executionDB, nil, err
+		}
 
-	if initFile != "" {
-		if err := extractSnapshot(initFile, stack.InstanceDir(), config.Init.ImportWasm); err != nil {
+		if chainConfig == nil {
+			chainConfig, err = getChainConfig(config, initDataReader, executionDB)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+
+		parsedInitMessage, err := getConsensusParsedInitMsg(ctx, config, chainId, l1Client, rollupAddrs, chainConfig)
+		if err != nil {
 			return nil, nil, err
 		}
+
+		l2BlockChain, chainConfig, err = getNewBlockchain(parsedInitMessage, config, initDataReader, chainConfig, genesisArbOSInit, executionDB, cacheConfig, tracer)
+		if err != nil {
+			return executionDB, nil, err
+		}
+
+		err = getAndValidateGenesisAssertion(ctx, config, l2BlockChain, initDataReader, &rollupAddrs, l1Client)
+		if err != nil {
+			return executionDB, nil, err
+		}
 	}
 
-	var initDataReader statetransfer.InitDataReader = nil
+	if chainConfig == nil {
+		return executionDB, nil, fmt.Errorf("chainConfig should not have been nil")
+	}
 
-	chainDB, err := stack.OpenDatabaseWithOptions("l2chaindata", node.DatabaseOptions{AncientsDirectory: config.Persistent.Ancient, MetricsNamespace: "l2chaindata/", Cache: config.Execution.Caching.DatabaseCache, Handles: config.Persistent.Handles, PebbleExtraOptions: persistentConfig.Pebble.ExtraOptions("l2chaindata")})
+	err = PruneExecutionDB(ctx, executionDB, stack, config, cacheConfig, persistentConfig, l1Client, rollupAddrs)
```

Comment:
OK, I didn't notice this before, but pruning executionDB requires consensusDB and the parent chain 😬, this is annoying.
Nothing to do in this PR regarding this BTW.

Reply:
Sounds good!